packages/material-ui/src/styles/withStyles.js

```diff
@@ -210,6 +213,62 @@ const withStyles = (stylesOrCreator, options = {}) => Component => {
       }
     }
 
+    getClasses() {
+      let generate = false;
+
+      if (!this.disableStylesGeneration) {
+        const sheetManager = this.sheetsManager.get(this.stylesCreatorSaved);
+        const sheetsManagerTheme = sheetManager.get(this.theme);
+        if (sheetsManagerTheme.sheet.classes !== this.lastRenderedClasses) {
+          this.lastRenderedClasses = sheetsManagerTheme.sheet.classes;
+          generate = true;
+        }
+      }
+
+      if (this.props.classes) {
+        if (this.props.classes !== this.lastClassesProp) {
+          this.lastClassesProp = this.props.classes;
+          generate = true;
+        }
+        if (generate) {
+          this.classes = {
+            ...this.lastRenderedClasses,
+            ...Object.keys(this.props.classes).reduce((accumulator, key) => {
+              warning(
+                this.lastRenderedClasses[key] || this.disableStylesGeneration,
+                [
+                  `Material-UI: the key \`${key}\` ` +
+                    `provided to the classes property is not implemented in ${getDisplayName(
+                      Component,
+                    )}.`,
+                  `You can only override one of the following: ${Object.keys(this.lastRenderedClasses).join(
+                    ',',
+                  )}`,
+                ].join('\n'),
+              );
+  
+              warning(
+                !this.props.classes[key] || typeof this.props.classes[key] === 'string',
+                [
+                  `Material-UI: the key \`${key}\` ` +
+                    `provided to the classes property is not valid for ${getDisplayName(Component)}.`,
+                  `You need to provide a non empty string instead of: ${this.props.classes[key]}.`,
+                ].join('\n'),
+              );
+  
+              if (this.props.classes[key]) {
+                accumulator[key] = `${this.lastRenderedClasses[key]} ${this.props.classes[key]}`;
+              }
+  
+              return accumulator;
+            }, {}),
```

Comment:
All the same logic as in the previous `render` method. The new object is generated and cached in `this.classes` such that future references to it don't generate new pointers.